### DISCUSSION TITLE
CIELAB -> RGB -> HSL

### DISCUSSION THREAD

**Discussion by blue-j** [colour-science/colour#1143] [Q&A]
What is the best choice of RGB color space in such a transform?  CIE RGB?  Rec.2020?

: )

J

**Comment by tjdcs** [comment]
Depends on what the RGB space is supposed to represent? RGB usually means it's some how display encoded (but not always) so you should pick an RGB model that matches your use case. Somewhere in Colour there is a long list of already implemented display models. 

On the other hand... I'm curious what your use case in HSL. If you are trying to perfrom perceptual image adjustments, you don't need to go to an RGB space yet. You can perform transforms / image tuning in an LCH representation of Lab directly. Or use an LCH model from something like CIECAM02

**Reply by blue-j** [reply]
Thank you so much for sharing your knowledge.  We are basically allowing the user to input CIELAB or RGB values in any bit depth, (along with illuminant or primary coordinates additionally, or a selected RGB space) and then converting the values to every color space / model known to Colour.  It is for industrial colorists or imaging professionals to reference.

In the CIELAB pipeline, I believed we had to convert to RGB before getting to a number of spaces:  CMY, CMYK, HSL, HSV, IHLS, Prismatic, RGB Luminance, YCbCr, YccBccRc, YCoCg.

So were we to use an RGB intermediate, I suppose it would need to shift the inputs minimally.  Here's the chunk of code where that happens:

```
            outputRGB = 'ITU-R BT.2020'
            wt = np.array(colour.RGB_COLOURSPACES[outputRGB].whitepoint)
            rd = np.array(colour.RGB_COLOURSPACES[outputRGB].primaries[0])
            gr = np.array(colour.RGB_COLOURSPACES[outputRGB].primaries[1])
            bl = np.array(colour.RGB_COLOURSPACES[outputRGB].primaries[2])
            # produce an array with the primaries xy coordinates
            p = np.array([rd, gr, bl])
            # produce the NPM using the primaries array and the white point xy coordinates of the input values
            matrix_RGB_to_XYZ = colour.normalised_primary_matrix(p, wt)
            matrix_XYZ_to_RGB = np.linalg.inv(matrix_RGB_to_XYZ)

            rgb_decoded = colour.XYZ_to_RGB(ciexyz, illuminant_XY, wt, matrix_XYZ_to_RGB)
```

So, we are going from CIELAB to XYZ to RGB to HSV, etc.

Sounds like there is another path?  Thanks again!

- J

**Comment by blue-j** [comment]
As an important side note, the code I shared does not actually seem to work.  I cannot get XYZ_to_RGB to work for the life of me!  The values I am seeing right now for CIEXYZ are [ 0.07216608  0.04415477  0.03858796] which is pristine and accurate for CIELAB ([25, 35, 5]).  Then the illuminant_XY (we are iterating through the illuminants, so for example, for illuminant A, this variable would hold [0.44758  0.40745].  illuminant_RGB/wt = [ 0.3127  0.329 ] for BT.2020.

the NPM is:

[[  6.36958048e-01   1.44616904e-01   1.68880975e-01]
 [  2.62700212e-01   6.77998072e-01   5.93017165e-02]
 [  4.99410657e-17   2.80726930e-02   1.06098506e+00]]

then the inverse of that is (matrix XYZ to RGB):

[[ 1.71665119 -0.35567078 -0.25336628]
 [-0.66668435  1.61648124  0.01576855]
 [ 0.01763986 -0.04277061  0.94210312]]

and the rgb_decoded is [ 0.15989773  0.07287098 -0.11008941].  No bueno.
